File: surf-chat/src/main/java/dev/slne/surf/social/chat/command/channel/ChannelListCommand.java
package dev.slne.surf.social.chat.command.channel;

import dev.jorel.commandapi.CommandAPICommand;
import dev.jorel.commandapi.arguments.IntegerArgument;
import dev.slne.surf.social.chat.object.Channel;
import dev.slne.surf.social.chat.provider.ChannelProvider;
import dev.slne.surf.social.chat.util.MessageBuilder;
import dev.slne.surf.social.chat.util.PageableMessageBuilder;
import net.kyori.adventure.text.Component;

public class ChannelListCommand extends CommandAPICommand {

  public ChannelListCommand(String commandName) {
    super(commandName);

    withOptionalArguments(new IntegerArgument("page"));
    executesPlayer((player, args ) -> {
      PageableMessageBuilder message = new PageableMessageBuilder();
      Integer page = args.getOrDefaultUnchecked("page", 1);

      int index = 0;

      message.setPageCommand("/channel list %page%");

      for (Channel channel : ChannelProvider.getInstance().getChannels().values()) {
        index ++;

        message.addLine(new MessageBuilder().darkSpacer(index + ". ").primary(channel.getName()).darkSpacer(" (").info(String.valueOf(channel.getMembers().size() + channel.getModerators().size() + 1)).darkSpacer(")").build().hoverEvent(this.createInfoMessage(channel)));
      }

      message.send(player, page);
    });
  }

  private Component createInfoMessage(Channel channel) {
    return new MessageBuilder()
        .primary("Kanalinformation: ").info(channel.getName()).newLine()
        .variableKey("Beschreibung: ").variableValue(channel.getDescription()).newLine()
        .variableKey("Besitzer: ").variableValue(channel.getOwner().getName()).newLine()
        .variableKey("Status: ").variableValue(channel.isClosed() ? "Geschlossen" : "Offen").newLine()
        .variableKey("Mitglieder: ").variableValue(String.valueOf(channel.getMembers().size() + channel.getModerators().size() + 1)).newLine()
        .variableKey("Einladungen: ").variableValue(String.valueOf(channel.getInvites().size())).newLine()
        .build();
  }
}

File: surf-chat/src/main/java/dev/slne/surf/social/chat/command/channel/ChannelListCommand.kt
package dev.slne.surf.social.chat.command.channel

import dev.jorel.commandapi.CommandAPICommand
import dev.jorel.commandapi.arguments.IntegerArgument
import dev.jorel.commandapi.executors.CommandArguments
import dev.jorel.commandapi.executors.PlayerCommandExecutor
import dev.jorel.commandapi.kotlindsl.integerArgument
import dev.jorel.commandapi.kotlindsl.playerExecutor
import dev.slne.surf.social.chat.`object`.Channel
import dev.slne.surf.social.chat.provider.ChannelProvider
import dev.slne.surf.social.chat.util.MessageBuilder
import dev.slne.surf.social.chat.util.PageableMessageBuilder
import net.kyori.adventure.text.Component
import org.bukkit.Bukkit
import org.bukkit.OfflinePlayer
import org.bukkit.entity.Player

class ChannelListCommand(commandName: String) : CommandAPICommand(commandName) {
    init {
        integerArgument("page", 1, Int.MAX_VALUE, true)
        playerExecutor { player, args ->
            val message = PageableMessageBuilder()
            val page = args.getOrDefaultUnchecked("page", 1)

            var index = 0

            message.setPageCommand("/channel list %page%")

            for (channel in ChannelProvider.channels.values) {
                index++

                message.addLine(
                    MessageBuilder().variableValue("$index. ").primary(channel.name)
                        .darkSpacer(" (")
                        .info((channel.members.size + channel.moderators.size + 1).toString())
                        .darkSpacer(")").build().hoverEvent(
                            this.createInfoMessage(channel)
                        )
                )
            }
            message.send(player, page)
        }
    }

    private fun createInfoMessage(channel: Channel): Component {
        val owner = channel.owner ?: return MessageBuilder().error("Ein Fehler ist aufgetreten.").build()
        val ownerPlayer = Bukkit.getOfflinePlayer(owner)
        return MessageBuilder()
            .primary("Kanalinformation: ").info(channel.name).newLine()
            .darkSpacer("   - ").variableKey("Beschreibung: ").variableValue(channel.description)
            .newLine()
            .darkSpacer("   - ").variableKey("Besitzer: ").variableValue(ownerPlayer.name ?: ownerPlayer.uniqueId.toString())
            .newLine()
            .darkSpacer("   - ").variableKey("Status: ")
            .variableValue(if (channel.closed) "Geschlossen" else "Offen").newLine()
            .darkSpacer("   - ").variableKey("Mitglieder: ")
            .variableValue((channel.members.size + channel.moderators.size + 1).toString())
            .newLine()
            .darkSpacer("   - ").variableKey("Einladungen: ")
            .variableValue(channel.invites.size.toString()).newLine()
            .build()
    }
}
